elftools/ehabi/decoder.py

```diff
@@ -0,0 +1,285 @@
+        opcode = self._bytecode_array[self._index]
+        self._index += 1
+        return 'pop %s' % self._printGPR((((1 << ((opcode & 0x7) + 1)) - 1) << 4))
+
+    def _decode_10101nnn(self):
+        # SW.startLine() << format("0x%02X      ; pop ", Opcode);
+        # PrintGPR((((1 << ((Opcode & 0x7) + 1)) - 1) << 4) | (1 << 14));
+        opcode = self._bytecode_array[self._index]
+        self._index += 1
+        return 'pop %s' % self._printGPR((((1 << ((opcode & 0x7) + 1)) - 1) << 4) | (1 << 14))
+
+    def _decode_10110000(self):
+        # SW.startLine() << format("0x%02X      ; finish\n", Opcode);
+        self._index += 1
+        return 'finish'
+
+    def _decode_10110001_0000iiii(self):
+        # SW.startLine()
+        #   << format("0x%02X 0x%02X ; %s", Opcode0, Opcode1,
+        #             ((Opcode1 & 0xf0) || Opcode1 == 0x00) ? "spare" : "pop ");
+        # if (((Opcode1 & 0xf0) == 0x00) && Opcode1)
+        #   PrintGPR((Opcode1 & 0x0f));
+        self._index += 1
+        op1 = self._bytecode_array[self._index]
+        self._index += 1
+        if (op1 & 0xf0) != 0 or op1 == 0x00:
+            return 'spare'
+        else:
+            return 'pop %s' % self._printGPR((op1 & 0x0f))
+
+    def _decode_10110010_uleb128(self):
+        #  SmallVector<uint8_t, 4> ULEB;
+        #  do { ULEB.push_back(Opcodes[OI ^ 3]); } while (Opcodes[OI++ ^ 3] & 0x80);
+        #  uint64_t Value = 0;
+        #  for (unsigned BI = 0, BE = ULEB.size(); BI != BE; ++BI)
+        #    Value = Value | ((ULEB[BI] & 0x7f) << (7 * BI));
+        #  OS << format("; vsp = vsp + %" PRIu64 "\n", 0x204 + (Value << 2));
+        self._index += 1
+        uleb_buffer = [self._bytecode_array[self._index]]
+        self._index += 1
+        while self._bytecode_array[self._index] & 0x80 == 0:
+            uleb_buffer.append(self._bytecode_array[self._index])
+            self._index += 1
+        value = 0
+        for b in reversed(uleb_buffer):
+            value = (value << 7) + (b & 0x7F)
+        return 'vsp = vsp + %u' % (0x204 + (value << 2))
+
+    def _decode_10110011_sssscccc(self):
+        # these two decoders are equal
+        return self._decode_11001001_sssscccc()
+
+    def _decode_101101nn(self):
+        return self._spare()
+
+    def _decode_10111nnn(self):
+        #  SW.startLine() << format("0x%02X      ; pop ", Opcode);
+        #  PrintRegisters((((1 << ((Opcode & 0x07) + 1)) - 1) << 8), "d");
+        opcode = self._bytecode_array[self._index]
+        self._index += 1
+        return 'pop %s' % self._print_registers((((1 << ((opcode & 0x07) + 1)) - 1) << 8), "d")
+
+    def _decode_11000110_sssscccc(self):
+        #  SW.startLine() << format("0x%02X 0x%02X ; pop ", Opcode0, Opcode1);
+        #  uint8_t Start = ((Opcode1 & 0xf0) >> 4);
+        #  uint8_t Count = ((Opcode1 & 0x0f) >> 0);
+        #  PrintRegisters((((1 << (Count + 1)) - 1) << Start), "wR");
+        self._index += 1
```

Comment:
In some `_decode`s you first grab a byte from `self._index` and then increment it, and in some you first increment it. Can you add a comment explaining the two scenarios?